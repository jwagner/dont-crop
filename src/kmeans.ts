// import mcgRandom from './mcgRandom';

type Point = {
  x: number;
  y: number;
  z: number;
};

export interface Cluster extends Point {
  count: number;
}

const channels = 3;
// sadly this is about twice as fast as x ** 2.
const squared = (x: number) => x * x;

// function selectClusterAtRandom(image: ImageData, random: () => number): Cluster {
//   const i = ((random() * image.width * image.height) | 0) * channels;
//   const { data } = image;
//   return {
//     x: data[i],
//     y: data[i + 1],
//     z: data[i + 2],
//     count: 0,
//   };
// }

// This function also operates without doing any gamma correction.
// I haven't evaluated the effect of this on the output.
// I could imagine that operating on the linear values would perform
// worse than the naive implementation here because the gamma encoding
// loosely correlates with the human perception.
// So the euclidian distance between the gamma encoded values might
// more closely match the perceived distance between the colors.
// Or it's just an excuse to save a few cycles. ¯\_(ツ)_/¯

export function kmeans(
  data: Float64Array,
  initialClusters: Point[],
  iterations: number,
): Cluster[] {
  // const { data } = image;
  const clusters: Cluster[] = [];
  const nextClusters: Cluster[] = [];
  // (histogramColors(image, k)).forEach(([r, g, b]) => {
  //   const { l: x, a: y, b: z } = srgbToLab({ r: r / 255, g: g / 255, b: b / 255 });
  //   clusters.push({
  //     x, y, z, count: 0,
  //   });
  //   nextClusters.push({
  //     x: 0, y: 0, z: 0, count: 0,
  //   });
  // });
  initialClusters.forEach((point) => {
    nextClusters.push({ ...point, count: 1 });
    clusters.push({
      x: 0, y: 0, z: 0, count: 0,
    });
  });
  for (let i = 0; i < iterations; i++) {
    clusters.forEach((cluster, j) => {
      // move next clusters to current clusters
      const nextCluster = nextClusters[j];
      const count = Math.max(nextCluster.count, 1);
      cluster.x = nextCluster.x / count;
      cluster.y = nextCluster.y / count;
      cluster.z = nextCluster.z / count;
      cluster.count = nextCluster.count;

      // reset next clusters
      nextCluster.x = 0;
      nextCluster.y = 0;
      nextCluster.z = 0;
      nextCluster.count = 0;
    });

    for (let p = 0; p < data.length; p += 3) {
      const x = data[p];
      const y = data[p + 1];
      const z = data[p + 2];
      let closestClusterDistance = Infinity;
      let closestCluster: Cluster = clusters[0];
      // find the closest cluster
      for (let ci = 0; ci < clusters.length; ci++) {
        // since this is just used in a comparison the sqrt is not needed
        const cluster = clusters[ci];
        const distance = squared(x - cluster.x)
          + squared(y - cluster.y)
          + squared(z - cluster.z);

        if (distance < closestClusterDistance) {
          closestClusterDistance = distance;
          closestCluster = nextClusters[ci];
        }
      }
      // add pixel to the closest cluster
      closestCluster.count += 1;
      closestCluster.x += x;
      closestCluster.y += y;
      closestCluster.z += z;
    }
  }
  return clusters;
}
