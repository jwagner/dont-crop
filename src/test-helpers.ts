import { join } from 'path';
import sharp from 'sharp';
import { srgbToLabComposed } from './lab';

export async function getTestImageData(): Promise<ImageData> {
  const src = join(__dirname, '../tests/image.jpg');
  const { data, info } = await sharp(src).ensureAlpha()
    .toColorspace('srgb')
    .raw()
    .toBuffer({ resolveWithObject: true });
  const imageData: ImageData = {
    width: info.width,
    height: info.height,
    data: new Uint8ClampedArray(data.buffer),
  };
  return imageData;
}

export async function getLabTestData() {
  const imageData = await getTestImageData();
  return srgbToLabComposed(imageData.data);
}
