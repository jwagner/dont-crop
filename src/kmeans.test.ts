import { kmeans } from './kmeans';
import { getLabTestData } from './test-helpers';

describe('kmeans', () => {
  const colors = [
    { x: 50, y: -32, z: 0 },
    { x: 60, y: -25, z: 2 },
    { x: 0, y: 0, z: 0 },
    { x: 100, y: 0, z: 0 },
  ];
  const components = ['x', 'y', 'z'] as const;
  const perColor = 4;
  const nPixels = colors.length * perColor;
  const pixels = new Float64Array(nPixels * components.length).map((_, i) => {
    const color = colors[((i / nPixels / components.length) * colors.length) | 0];
    const component = components[i % (components.length)];
    return color[component];
  });
  test('perfect initial clusters', () => {
    const clusters = kmeans(pixels, colors, 8);
    colors.forEach((color, i) => {
      expect(clusters[i].count).toEqual(perColor);
      expect(clusters[i].x).toBeCloseTo(color.x);
      expect(clusters[i].y).toBeCloseTo(color.y);
      expect(clusters[i].z).toBeCloseTo(color.z);
    });
  });
  test('randomized initial clusters', () => {
    const random = 4; // chosen by fair dice roll.
    // guaranteed to be random.
    const inital = colors.map((c) => ({
      x: c.x + random,
      y: c.y + random,
      z: c.z + random,
    }));
    const clusters = kmeans(pixels, inital, 16);
    colors.forEach((color, i) => {
      expect(clusters[i].count).toEqual(perColor);
      expect(clusters[i].x).toBeCloseTo(color.x);
      expect(clusters[i].y).toBeCloseTo(color.y);
      expect(clusters[i].z).toBeCloseTo(color.z);
    });
  });
  test('fewer clusters than colors', () => {
    const random = 4; // chosen by fair dice roll.
    // guaranteed to be random.
    const inital = colors.map((c) => ({
      x: c.x + random,
      y: c.y + random,
      z: c.z + random,
    })).slice(0, 3);
    const clusters = kmeans(pixels, inital, 16);
    expect(clusters).toMatchInlineSnapshot(`
[
  {
    "count": 8,
    "x": 55,
    "y": -28.5,
    "z": 1,
  },
  {
    "count": 4,
    "x": 100,
    "y": 0,
    "z": 0,
  },
  {
    "count": 4,
    "x": 0,
    "y": 0,
    "z": 0,
  },
]
`);
  });
  test('1x1 image', () => {
    const result = kmeans(pixels.subarray(0, 3), colors, 4);
    const zeros = result.filter((c) => c.count === 0);
    const nonZero = result.filter((c) => c.count > 0);
    expect(zeros).toHaveLength(3);
    expect(nonZero).toEqual([{
      count: 1,
      x: 50,
      y: -32,
      z: 0,
    }]);
  });
  test('matches snapshot', async () => {
    const data = await getLabTestData();
    const initial = [
      { x: 0, y: 0, z: 0 },
      { x: 50, y: 0, z: 0 },
      { x: 100, y: 0, z: 0 },
    ];
    const actual = kmeans(data, initial, 4);
    expect(actual).toMatchSnapshot();
  });
});
