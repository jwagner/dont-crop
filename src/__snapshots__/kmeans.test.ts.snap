// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kmeans matches snapshot 1`] = `
Array [
  Object {
    "count": 2731,
    "x": 13.589797352269795,
    "y": 2.4418455704663105,
    "z": -0.9805334640561165,
  },
  Object {
    "count": 4803,
    "x": 56.53537330534744,
    "y": 42.42230747663517,
    "z": 45.737966169693614,
  },
  Object {
    "count": 3346,
    "x": 76.60102753535544,
    "y": 0.6589080592547374,
    "z": 14.062375232359388,
  },
]
`;
